### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by eduardodelreal**
Buenas noches Guille. Subo mi análisis. Tengo varias dudas que me gustaría compartir contigo:
- está bien hecho el approach de utilizar SMOTE para balancear más las clases o no debo hacerlo? por ejemplo, al usarlo en el modelo SVM con linear kernel, se ha sobreajustado, aunque he sabido detectarlo ya que el resultado era perfecto (lo he dejado para que le eches un ojo).
- al principio, he hecho un randomForest para ver que features contribuyen o no. Había un par de ellas que su peso era cero. Significa que debo quitarlas de la X_train para los subsiguientes modelos? Yo he asumido que no, ya que alomejor con otro modelo si que coge relevancia, pero no estoy seguro.
- Mis resultados son bastantes parejos entre varios modelos de clasificacion. Me he basado en el AUC, el area de la precision recall y la sensitivity y el recall que se da como resultado. Lo he valorado bien? Como podría decantarme por un modelo ante situaciones asi?

Gracias y un saludo.

### CODE REVIEW

**Comment by guillerbf** [comment]
Firstly, I don't think this branch has been created from where it should have since it includes tons of code from module 1 and 2 that shouldn't be here.

A PR should only contain the code required for the given piece of work you are submitting for review. Nothing else.

File: src/module_3/module_3_zrive.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
+# Filter orders with more than five products ordered
+#filtered_data = data[data['user_order_seq'] > 5]
+
+# Save the filtered dataset to a new CSV file
+#filtered_data.to_csv("filtered_feature_frames.csv", index=False)
+def main():
+    print(data)
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.py:47]: Why are you uploading this file for review?

File: src/module_3/module_3_zrive.md [hunk 2]
```diff
@@ -0,0 +1,1525 @@
...
+
+First of all, let's load the dataset and filter it so that we can get only those orders which have 5 or more products on it:
+
+
+
+```python
+data=pd.read_csv("feature_frames.csv")
+filtered_data = data[data['user_order_seq'] > 5]
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+logging.info("Las primeras filas del DataFrame son:\n%s", filtered_data.head().to_string())
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:18]: Keep it in english... 😅

File: src/module_3/module_3_zrive.md [hunk 3]
```diff
@@ -0,0 +1,1525 @@
...
+import numpy as np
+```
+
+First of all, let's load the dataset and filter it so that we can get only those orders which have 5 or more products on it:
+
+
+
+```python
+data=pd.read_csv("feature_frames.csv")
+filtered_data = data[data['user_order_seq'] > 5]
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:16]: Are you sure this is filtering orders with 5+ products? 

This is removing the 5 first orders for every user! A nice thing when filtering is to look at sizes before/after. This should leave a very very small dataset.

File: src/module_3/module_3_zrive.md [hunk 4]
```diff
@@ -0,0 +1,1525 @@
...
+
+
+df['season'] = df['order_date'].dt.month.apply(get_season)
+
+df['day_of_week'] = df['order_date'].dt.dayofweek
+df['is_weekend'] = df['order_date'].dt.dayofweek >= 5
+df['hour_of_day'] = df['order_date'].dt.hour
+
+X = df.drop(['variant_id', 'order_id', 'user_id', 'created_at', 'order_date', 'outcome'], axis=1)
+X = pd.concat([X, df[['day_of_week', 'is_weekend', 'hour_of_day']]], axis=1)
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:389]: Have you run this cell in order? You shouldn't need to do concat since those columns have already being assigned to `df` and aren't included in `drop`

File: src/module_3/module_3_zrive.md [hunk 5]
```diff
@@ -0,0 +1,1525 @@
...
+    27  ricepastapulses               6             0.0               0.0             0.0             0.0          0.081052      0.053512  clearspring           0.039683           2.0             0.0           0.0         0.0             2.0                            33.0                        42.0                   31.134053                              30.0                          30.0                      24.27618       4            3       False            0            3       False            0
+
+
+Now that we have divided our X & y ; let's try to select among the variables we have, which ones are the most importants. For this purpose, I will use the Random Forest Classifier.
+
+
+```python
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import train_test_split
+X_encoded = pd.get_dummies(X)
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:411]: This is super error prone, since you don't control which variables will be one-hot encoded (it depends on the type assumed by pandas). Moreover, we saw in module 2 that some categorical variables had huge cardinality. 

Moreover, if you do one-hot encoding to select then features, it could be the case that some dimensions of a one-hot encoded variable are eliminated, which may not be the best idea.

File: src/module_3/module_3_zrive.md [hunk 6]
```diff
@@ -0,0 +1,1525 @@
...
+    27  ricepastapulses               6             0.0               0.0             0.0             0.0          0.081052      0.053512  clearspring           0.039683           2.0             0.0           0.0         0.0             2.0                            33.0                        42.0                   31.134053                              30.0                          30.0                      24.27618       4            3       False            0            3       False            0
+
+
+Now that we have divided our X & y ; let's try to select among the variables we have, which ones are the most importants. For this purpose, I will use the Random Forest Classifier.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:405]: You can do feature selection using linear models (ie. lasso) why going straight to random forest?

File: src/module_3/module_3_zrive.md [hunk 7]
```diff
@@ -0,0 +1,1525 @@
...
+    avg_days_to_buy_product_type                          0.838103  
+    std_days_to_buy_product_type                          1.000000  
+    
+    [23 rows x 23 columns]
+
+
+
+    
+![png](module_3_zrive_files/module_3_zrive_4_1.png)
+
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:359]: You have done correlations but do not comment anything on it. Are you thinking about using high-correlated features to reduce dimensionality? Aren't they correlated enough? Do you leave it here to come back later based on other modelling steps?

File: src/module_3/module_3_zrive.md [hunk 8]
```diff
@@ -0,0 +1,1525 @@
...
+        return 1  # winter
+    elif month in [3, 4, 5]:
+        return 2  # spring
+    elif month in [6, 7, 8]:
+        return 3  # summer
+    else:
+        return 4  # Outumn
+
+
+df['season'] = df['order_date'].dt.month.apply(get_season)
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:382]: You are assigning numbers (therefore creating in fact an ordinal variables) but, is there any order in seasons? Should it be a categorical variable instead subject to be encoded in different ways?

File: src/module_3/module_3_zrive.md [hunk 9]
```diff
@@ -0,0 +1,1525 @@
...
+             0.0       0.99      1.00      0.99     58698
+             1.0       0.68      0.03      0.06       835
+    
+        accuracy                           0.99     59533
+       macro avg       0.84      0.52      0.53     59533
+    weighted avg       0.98      0.99      0.98     59533
+    
+
+
+The Logistic Regression model achieves 98.62% accuracy, indicating high overall performance. However, it excels at identifying the majority class (0.0) with nearly 99% precision and 100% recall but struggles with the minority class (1.0), showing only 69% precision and a mere 3% recall. This discrepancy suggests difficulties in detecting the less frequent class, likely due to class imbalance. Macro averages reveal the challenge with a lower performance on the minority class.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:843]: We discussed it last week, accuracy should never be used if the problem is imbalanced!! 

You can predict almost everything as "0" and still get a pretty decent accuracy

File: src/module_3/module_3_zrive.md [hunk 10]
```diff
@@ -0,0 +1,1525 @@
...
+
+```
+
+
+    
+![png](module_3_zrive_files/module_3_zrive_16_0.png)
+    
+
+
+The ROC curve indicates a binary classifier's performance, with an AUC of 0.78 suggesting good discriminative ability. The curve stays well above the line of randomness (diagonal dashed line, AUC = 0.5), indicating that the classifier is useful. The graph implies that as the threshold is varied, the classifier achieves a reasonable balance between sensitivity and specificity. The model is notably better than a random guess but not perfect, with some room for improvement.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:908]: Careful here, False Positive Rate isn't specificity!

Comment (guillerbf) [inline src/module_3/module_3_zrive.md:908]: ROC curve way more useful than looking at accuracies.

File: src/module_3/module_3_zrive.md (hunk 9 above)
Comment (guillerbf) [reply src/module_3/module_3_zrive.md:843]: Moreover, remember that this confusion matrix will be calculated by default assuming a decision threshold of 0.5, which with such an imbalance problem will be most likely not optimal at all.

File: src/module_3/module_3_zrive.md [hunk 11]
```diff
@@ -0,0 +1,1525 @@
...
+             0.0       0.99      0.84      0.91     58698
+             1.0       0.06      0.66      0.10       835
+    
+        accuracy                           0.84     59533
+       macro avg       0.53      0.75      0.51     59533
+    weighted avg       0.98      0.84      0.90     59533
+    
+
+
+Now, the accuracy has decreased but there's a bigger recall for the minority class
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:876]: Now it looks somewhat better because as you "artificially" balance the problema, a threshold of 0.5 makes more sense. Still, it's arbitrary and likely still suboptimal. That's why I suggest looking at ROC and precision-recall rather than confusion matrix.

Once we have decided the working threshold based on those curves we can come back an print the confusion matrix.

File: src/module_3/module_3_zrive.md [hunk 12]
```diff
@@ -0,0 +1,1525 @@
...
+
+
+    
+![png](module_3_zrive_files/module_3_zrive_19_0.png)
+    
+
+
+This curve shows a model with high initial precision and very low recall, indicating accurate but infrequent positive predictions. As recall increases, precision drops sharply, suggesting the model incorrectly predicts more negatives as positives. The low area under the curve (AUC = 0.12) implies overall poor model performance in classifying the positive class in a binary classification task.
+
+The model is biased toward the majority class, identifying most of its instances correctly but at the cost of misclassifying many instances of the minority class as the majority class.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:935]: It's not the model that is biased, is that the dataset is biased due to its imbalance.

File: src/module_3/module_3_zrive.md [hunk 13]
```diff
@@ -0,0 +1,1525 @@
...
+The low precision but high recall for the minority class indicates the model is overpredicting the minority class, resulting in a high number of false positives.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:936]: That's what you would expect in this type of scenarios. That's why it's so important to look at precision-recall since they only put attention on the minority class, which most of the times is the class of interest.

File: src/module_3/module_3_zrive.md [hunk 14]
```diff
@@ -0,0 +1,1525 @@
...
+
+
+LET'S TRY SOME OTHER LINEAR MODELS. I'M GOING TO DIRECTLY AVOID USING LINEAR REGRESSION SINCE THE OUTPUT I WANT TO VERIFY IS NOT CONTINUOUS. BUT I'M GOING TO DO THE FOLLOWING LINEAR MODELS:  LDA, SVM WITH LINEAR KERNEL, STOCHASTIC GRADIENT DESCENT & GENERALIZED LINEAR MODELS. AFTER DOING SO, I AM GOING TO EVALUATE THE BEST MODEL FOR OUR GOAL USING EXPERIMENTATION OR CROSS VALIDATION.
+
+UP TO NOW, OPTING FOR LINEAR REGRESSION OVER LASSO HINGES ON VALUING SIMPLICITY AND CLEAR INTERPRETABILITY. LINEAR REGRESSION'S DIRECT COEFFICIENT INTERPRETATION OFFERS INSIGHTS INTO HOW PREDICTORS IMPACT THE OUTCOME, IDEAL WHEN ALL VARIABLES ARE RELEVANT AND THE GOAL IS TO UNDERSTAND THEIR INFLUENCE WITHOUT THE COMPLEXITY OF LASSO'S REGULARIZATION. THIS APPROACH IS PREFERABLE IN LOW-DIMENSIONAL DATA WHERE MODEL TRANSPARENCY AND EASE OF EXPLANATION TO STAKEHOLDERS ARE KEY, MAKING IT A STRAIGHTFORWARD CHOICE FOR PROJECTS EMPHASIZING THE ELUCIDATION OF PREDICTOR EFFECTS.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:941]: This is not true. Lasso is as interpretable as linear regression. You could have also use it to see if some features where set to zero as an attempt to do feature importance (you will see more of this in module 5).

File: src/module_3/module_3_zrive.md [hunk 15]
```diff
@@ -0,0 +1,1525 @@
...
+    
+
+
+The SGD Classifier achieved an overall accuracy of 98.42%, demonstrating strong effectiveness in classifying the dataset, particularly for the majority class (0.0) with almost perfect precision and recall. However, its performance on the minority class (1.0) is weaker, with low precision and even lower recall, indicating challenges in accurately identifying responsive users. The low F1-score for this class highlights a poor balance between precision and recall, further evidenced by a significant number of false negatives (749) in the confusion matrix. This disparity suggests that while highly effective for non-responsive users, the model's ability to pinpoint responsive individuals needs refinement.
+
+
+
+
+
+CLEARLY, WE HAVE SEEN ACROSS ALL OF THESE DIFFERENT MODELS THE SAME PROBLEM, WHICH IS THE CLASS IMBALANCE.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:1171]: You have compared all the models using a accuracy, which is WRONG. I much better options would have been to plot the precision-recall curves for all the models together to see which one is better.

File: src/module_3/module_3_zrive.md [hunk 16]
```diff
@@ -0,0 +1,1525 @@
...
+
+A class imbalance occurs in classification problems when the target classes are disproportionately represented in the data set, that is, when the number of samples belonging to one class is much larger than the number of samples in other class.
+
+Therefore, we are going to repeat the approaches for all of these models, by using the SMOTE technique.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:1175]: SMOTE is an upsampling technique. You could have use a subsampling technique too. We can discuss it on Sunday!

File: src/module_3/module_3_zrive.md [hunk 17]
```diff
@@ -0,0 +1,1525 @@
...
+    
+
+
+
+    
+![png](module_3_zrive_files/module_3_zrive_40_2.png)
+    
+
+
+As we can see, the results are almost perfect. This is indeed not good, since the model has learned from the noise. The application of SMOTE in our linear SVM model has led to concerns of overfitting, as indicated by the almost perfect performance metrics that do not realistically reflect the complexities of unseen data. This over-optimistic result suggests that the model may not generalize well when exposed to new, real-world datasets, leading us to reconsider the use of SMOTE in this context. Ultimately, the high performance on the training set, coupled with concerns of generalization, signals that SMOTE may not be the most appropriate approach for our current modeling needs when talking of SVM linear.
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:1523]: You have fitted this one to a random dataset from sklearn...

File: src/module_3/module_3_zrive.md [hunk 18]
```diff
@@ -0,0 +1,1525 @@
...
+
+
+Now that we have divided our X & y ; let's try to select among the variables we have, which ones are the most importants. For this purpose, I will use the Random Forest Classifier.
+
+
+```python
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import train_test_split
+X_encoded = pd.get_dummies(X)
+X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)
```
Comment (guillerbf) [inline src/module_3/module_3_zrive.md:412]: Does it make sense to a random split? Shouldn't all the products of the same order be either all in train or all in test? Isn't there a temporal component to the problem to be taken into account? Is there any information bias in knowing that a user has bought a given product in a future order to predict if they will buy it in the one I'm predicting?

**Review by guillerbf — COMMENTED** [review]
My overall feeling is that the PR feels somewhat rushed like the last model trained over a random dataset or filtering the dataset with a wrong approach.

I would encourage you to take time to think about every step you take and every decision you make. It's not to get freeze but to make good informed decisions. Also, you have to try to avoid make wrong statements. An "I don't know" is way less dangerous than a "wrong and confident idea". 

Models aren't typically the main differentiation factor, comparing dozens of them is irrelevant if you are using the wrong metric to compare them. As we discussed, how we split the data and how measure success are the two most importante steps. Then, what data I feed the model. Finally, the model choice and model tuning.

Finally, regarding model tuning you have compared several models but haven't apply any regularisation. Linear models are less likely to overfit but still shouldn't be given for granted that they aren't. I'd at least check it.

The code itself is good. It's clean and well structured (for a notebook). We need to work on take a further step and start making good design decisions along the process.

**Comment by eduardodelreal** [comment]
Guille la seleccion de mi modelo no lo he hecho en base a la accuracy sino
en base a la AUC  de la ROC y la precision recall, simplemente menciono la
accuracy en muchos resultados porque esta ahi y por no olvidarla, pero no
lo he valorado en funcion de la accuracy
